### PR TITLE
[SCHEDULE] New Reduction Mode for Tensorize

### PR DESCRIPTION
**Pull request by kun-zh**
…e 714. @tqchen , This fix can resolve the original issue which I raised. Please review it , thanks.

### CODE REVIEW

File: src/op/op_util.cc [hunk 1]
```diff
@@ -208,5 +208,31 @@ Stmt Substitute(Stmt s,
   return ir::Substitute(s, init);
 }
 
+Stmt TransformUpdate(const Stage& stage,
+                     const std::unordered_map<IterVar, Range>& dom_map,
+                     Stmt body,
+                     Stmt update)
+{
+   Expr condition = (make_zero(Int(32)) ==  make_zero(Int(32))); //will try Bool(1)
```
Comment (tqchen) [inline src/op/op_util.cc]: build a ```Array<Expr>``` containing condition,  use https://github.com/dmlc/tvm/blob/master/src/arithmetic/compute_expr.h#L42 to do reduction

File: src/op/op_util.cc [hunk 2]
```diff
@@ -208,5 +208,31 @@ Stmt Substitute(Stmt s,
...
+      const IterVarAttr& attr = (*iit).second;
+      if (attr->iter_type == kTensorized) {
+        break;
+      }
+    }
+    if (iv->iter_type == kCommReduce) {
+      auto vit = dom_map.find(iv);
+      CHECK(vit != dom_map.end());
+      const Range& vrange = vit->second;
+      Expr newcond = ( iv->var == vrange->min);
```
Comment (tqchen) [inline src/op/op_util.cc]: use condition
```
if (likely(var1> min1) or likely(var2 > min2)) {
   update
} else {
   body
}
```

File: src/op/op_util.cc [hunk 3]
```diff
@@ -208,5 +208,31 @@ Stmt Substitute(Stmt s,
   return ir::Substitute(s, init);
 }
 
+Stmt TransformUpdate(const Stage& stage,
```
Comment (tqchen) [inline src/op/op_util.cc]: This function is specific to Tensorize, move it to tensorize.cc, no need to keep comment in header file, move comment document to tensorize.cc as well

File: src/op/tensorize.cc [hunk 4]
```diff
@@ -416,32 +416,48 @@ Stmt MakeTensorize(const ComputeOpNode* self,
...
+      // The update
+      Stmt update = MergeNest(output_bind_nest, intrin->reduce_update);
+      update = MergeNest(input_bind_nest, update);
+      update = Substitute(update, vmap);
+      update = MergeNest(binder.asserts(), update);
+      update = Substitute(update, n.main_vmap);
+      update = MergeNest(update_nest, update);
+      return MergeNest(common, Block::make(init, update));
+    } else {
+      // The update
```
Comment (tqchen) [inline src/op/tensorize.cc]: Add a comment about what this logic is about. Something like 
```
// when init intrinsic is not available, transform to reset in the first iter.
```

File: src/op/tensorize.cc [hunk 5]
```diff
@@ -416,32 +416,48 @@ Stmt MakeTensorize(const ComputeOpNode* self,
...
+      Stmt update = TransformUpdate (stage, dom_map,
```
Comment (tqchen) [inline src/op/tensorize.cc]: add CHECK(body.defined());

**Comment by tqchen** [comment]
Please add a unittestcase to https://github.com/dmlc/tvm/blob/master/tests/python/unittest/test_schedule_tensorize.py#L1

**Comment by tqchen** [comment]
there is one additional drawback of the approach which requires additional verification. We cannot let the ```n.main_predicates``` do not contain conditions on the outer Reduce IterVar. 

For example, if condition contains rc>1 (and rc is from [0, n) ), then the initial checking condition is no longer correct. Error shall be raised when such case is detected.

This can be done via something similar to https://github.com/dmlc/tvm/blob/master/src/op/tensorize.cc#L124

File: src/op/tensorize.cc [hunk 6]
```diff
@@ -416,32 +416,48 @@ Stmt MakeTensorize(const ComputeOpNode* self,
     return MergeNest(nest, body);
   } else {
     // Need to split reduction
-    CHECK(intrin->reduce_init.defined())
-        << "Reduction init op for intrin " << intrin << " is not defined";
+    // Comment out the following check for the case when there is no init func
+    //CHECK(intrin->reduce_init.defined())
```
Comment (tqchen) [inline src/op/tensorize.cc]: remove lgtm, simply remove them

**Comment by kun-zh** [comment]
Refined the code per review comment, and add a test case in unittest, will add more cases in the following days.

File: src/op/op_util.cc (hunk 3 above)
Comment (kun-zh) [reply src/op/op_util.cc]: done, move the function into tensorize.cc

File: src/op/op_util.cc (hunk 1 above)
Comment (kun-zh) [reply src/op/op_util.cc]: done, please check

File: src/op/op_util.cc (hunk 2 above)
Comment (kun-zh) [reply src/op/op_util.cc]: done

File: src/op/tensorize.cc (hunk 6 above)
Comment (kun-zh) [reply src/op/tensorize.cc]: done

File: src/op/tensorize.cc (hunk 4 above)
Comment (kun-zh) [reply src/op/tensorize.cc]: done

File: src/op/tensorize.cc (hunk 5 above)
Comment (kun-zh) [reply src/op/tensorize.cc]: done

File: tests/python/unittest/test_schedule_tensorize_init_none.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
+import tvm
+
```
Comment (tqchen) [inline tests/python/unittest/test_schedule_tensorize_init_none.py]: move this to test_schedule_tensorize.py We can use different function name

**Pull request merged**